### ISSUE TITLE
talisman --scan stuck due to blocked write

### ISSUE THREAD

**Issue by teleivo** [thoughtworks/talisman#301]
**Describe the bug**

`talisman --scan` intermittently but frequently hangs on a `git ls-tree -r` which is executed to gather all blobs and the commits they were in. This issue will appear before any tests on the file contents are executed. So I am only seeing the `Running Scan` ACSII art and nothing else.

Here is where the command is executed                                                                                                                                       
https://github.com/thoughtworks/talisman/blob/f6a09e119a1ddae1b3773c3204a1ebe1250e10f0/scanner/scanner.go#L45

This shows that a goroutine is created per commit to gather the tree and its blobs
https://github.com/thoughtworks/talisman/blob/f6a09e119a1ddae1b3773c3204a1ebe1250e10f0/scanner/scanner.go#L33-L40

**To Reproduce**

To reproduce this issue you might simply need to run `talisman --scan` on a                                                                                                 
repo like https://github.com/apache/bookkeeper. Make sure to run it multiple times. For me 6/10 times it will get stuck, soon after I start the scan.

If that does not cause the scan to hang try running the code in this repo https://github.com/teleivo/talisman-issue-blocked-write
I isolated the code causing this so it can easily be reproduced.

How do you know talisman is stuck?

This issue will appear before any tests on the file contents are executed. So
you will only see the `Running Scan` ACSII art and nothing else. You can also
do a

```sh
ps --ppid $(pgrep talisman) -F

 PID  PPID  C    SZ   RSS PSR STIME TTY          TIME CMD
4149  3874  0 18113     0   0 Apr22 pts/1    00:00:00 git ls-tree -r c9dc301feb48ca170c3d6205a36fca63a4950c5a
4162  3874  0 18113     0   2 Apr22 pts/1    00:00:00 git ls-tree -r 33ea58027b0a3ba160f7ac19d20568709f453f4d
```

You should also see one or more `git ls-tree -r` commands that just never
finish.

**Expected behavior**
talisman --scan should execute successfully and create a report.

**Desktop**
 - OS: Debian
 - Version: 10

**Additional context**

I isolated the code causing this so it can easily be reproduced in https://github.com/teleivo/talisman-issue-blocked-write
The repo also contains a detailed analysis of how I arrived at the root cause.

_Root cause_

The `git ls-tree` command that is executed writes bytes to STDOUT. The Go
[exec.Command().CombinedOutput()](https://golang.org/pkg/os/exec/#Cmd.CombinedOutput)
is waiting for this command to finish. The syscall executing the write to
STDOUT for `git ls-tree` gets stuck as the pipe buffer is full. Subsequent
write() calls will therefore block until the pipe buffer is read from
(drained). All this happens on the OS level, not visible to the Go
exec.Command. At least as far as I known.

strace shows me the syscalls for these stuck git processes. Here is a sample

```sh
strace -s 99 -ffp 4149
strace: Process 4149 attached
write(1, "AuditorElector.java\n100644 blob 9830c592904cf4848d4068b64e562b78e815b5dc\tbookkeeper-server/src/main"..., 4096
```

The syscall write specifies in its documentation that it will block if the buffer it uses is full. The process is trying to write to STDOUT (filedescriptor 1, given as the first argument). There is a pipe buffer in between which has a default capacity of 65,536 bytes.

These processes have written or want to write the maximum
amount of bytes:

```sh                                                                                                                                                                       
cat /proc/4162/io
rchar: 14488
wchar: 65536
syscr: 33
syscw: 36
read_bytes: 1634304
write_bytes: 0
cancelled_write_bytes: 0
```

```sh
cat /proc/4149/io
rchar: 14488
wchar: 65536
syscr: 33
syscw: 20
read_bytes: 491520
write_bytes: 0
cancelled_write_bytes: 0
```

https://git.kernel.org/pub/scm/linux/kernel/git/torvalds/linux.git/tree/Documentation/filesystems/proc.rst?id=HEAD#n1730

(assuming 1 byte per character as the git ls-tree output looks like ASCII at
least for these particular trees).

So the processes block on write as the pipe buffers are full.

Related documentation to read:

- man 2 write (tells you that write will block on a full pipe buffer)
- man 7 pipe (tells you the default pipe buffer size)

This line is where the command is executed
https://github.com/thoughtworks/talisman/blob/f6a09e119a1ddae1b3773c3204a1ebe1250e10f0/scanner/scanner.go#L45

_Interesting Correlation With Repository Sizes:_

I was wondering why I can scan the talisman repo without this ever occurring. I suspect its related to tree sizes/repo structures. I can easily reproduce it on Java repos. Java usually has deeper directory structures, longer package/class names than go.

Commits in https://github.com/apache/bookkeeper it gets stuck on:
de9eecc180cc78fb65b509d7ae0e5ea82fd5dc8c
44b1069e20ce6cc84302ed685936080f612bd8f0
f1a1f12195d8538338d000dba71fb2e98de0afbb
280a5cd685fce06a4f3d78ba28aa3cdf55084555
5311e00f399ae3b5cb39fe03d8b5dd8c0547de14
12525717adf45f9c32c12cfd6228935e98b2b51b
c11baba7dcedc227849021d0aee8303cfcc9765c
37e562225c83c227e5994e5a327181fca22d515b

and their respective sizes in bytes:
95213
446762
446667
2294198
413450
428426
440095
433768

Note that they are all > than the default pipe buffer of 65k. The scanner is able to gather trees larger than the default buffer size tough. It just happens that if it gets stuck I do see trees of a larger size.

I am not sure that the number of commits plays a role. The talisman repo only has 456 commits which is not a lot. I can reproduce the issue in https://github.com/teleivo/talisman-issue-blocked-write with that number of commits. The tree size has to be bigger than what it is in talisman (which is only 8k for the latest commit).

_Some Potential Solutions_
* gather blobs sequentially
* limit number of concurrent `git ls-tree -r` operations
* use an https://golang.org/pkg/io/#Pipe to stream from STDOUT directly instead of letting [exec.Command().CombinedOutput()](https://golang.org/pkg/os/exec/#Cmd.CombinedOutput) do it
* find an alternative of gathering blobs via a dependency like git2go or gogit instead of executing a command on the command line

I already have a working prototype which I can provide a draft PR for if that issue is reproduced by someone else 😄

**Comment by svishwanath-tw** [comment]
Thanks for this very detailed issue description @teleivo 


On reading https://unix.stackexchange.com/questions/11946/how-big-is-the-pipe-buffer i found out that
1. after kernel 2.6.11 all pipe are limited to 16 pages and if page-size is 4096 bytes, this will be 65535
2. darwin/bsd xnu starts with a limit of 16348 bytes and extends up to 65535 based on large writes

I've spent some time trying to re-create issue on my Mac (MBP 2018/16G/i7) using https://github.com/teleivo/talisman-issue-blocked-write, and could not re-create the failure after multiple runs.

Note: The `status=none` argument wasn't recognized on MacOS
I was unable to re-create the hanging with `bs=4096 count=1` even after bumping to `bs=66666 count=2`, it didn't fail.
I did notice that the process did seem to be unresponsive for about 2 seconds.

About using git2go, it would require a significant change to the code-base, 
We did look at gogit in the past and it lacks some of the features of git cli that we currently use
